server/src/main/java/org/opensearch/tasks/Task.java

```diff
@@ -195,6 +213,64 @@ public Status getStatus() {
         return null;
     }
 
+    /**
+     * Returns total resource consumption of the task
+     */
+    public Map<String, List<TaskCompleteResourceInfo>> getResourceStats() {
+        return resourceStats;
+    }
+
+    public Map<String, Long> getTotalResourceStats() {
+        return new HashMap<String, Long>() {
+            {
+                for (TaskStats taskStats : TaskStats.values()) {
+                    put(taskStats.toString(), getTotalResourceUtilization(taskStats));
+                }
+            }
+        };
+    }
+
+    /**
+     * Returns total resource consumption for a specific task stats.
+     */
+    public long getTotalResourceUtilization(TaskStats taskStats) {
+        long totalResourceConsumption = 0L;
+        for (List<TaskCompleteResourceInfo> resourceInfoList : resourceStats.values()) {
+            for (TaskCompleteResourceInfo resourceInfo : resourceInfoList) {
+                for (Map.Entry<TaskStatsType, TaskResourceInfo> entry : resourceInfo.getResourceInfo().entrySet()) {
+                    if (!entry.getKey().isOnlyForAnalysis()) {
+                        totalResourceConsumption += entry.getValue().getStatsInfo().get(taskStats).getTotalValue();
+                    }
+                }
+            }
+        }
+        return totalResourceConsumption;
+    }
+
+    /**
+     * Add or update thread's resource consumption stats
+     * @param threadId: threadId
+     * @param isActive: if the thread is currently active
+     * @param statsType: stats type
+     * @param taskStatsInfo: task resource metrics
+     */
+    public void addOrUpdateTaskResourceStats(String threadId, boolean isActive, TaskStatsType statsType, TaskStatsInfo... taskStatsInfo) {
+        List<TaskCompleteResourceInfo> resourceInfoList = resourceStats.getOrDefault(threadId, new ArrayList<>());
+        if (resourceInfoList.isEmpty()) {
+            resourceInfoList.add(new TaskCompleteResourceInfo(isActive, statsType, taskStatsInfo));
+            resourceStats.put(threadId, resourceInfoList);
+        } else {
+            for (TaskCompleteResourceInfo resourceInfo : resourceInfoList) {
+                // If present, update the exising active entry.
+                if (resourceInfo.isActive()) {
+                    resourceInfo.update(isActive, statsType, taskStatsInfo);
+                    return;
+                }
+            }
+            resourceInfoList.add(new TaskCompleteResourceInfo(isActive, statsType, taskStatsInfo));
```

Comment:
The `resourceStats` map is thread safe because it is a ConcurrentHashMap, but this code is not thread safe. I _think_ the assumption here is that there will never be concurrent execution from multiple threads for the same threadId, which means this code will be correct because the only concurrency to worry about is adding new items to the map. Is that right?

Reply:
Thats right. The assumption is that will never have concurrent execution from multiple threads for the same threadId.